API.md

```diff
@@ -0,0 +1,115 @@
+    }
+    
+    var okButton: UIButton {
+       return shared { return UIButton() }
+    }
+}
+```
+**Note:** It's up to you to decide what items make sense on the *DI Graph* and which items can be just local properties in your `ViewController` subclass. Anything that you'd like to mock during a test needs to be passed in (as a protocol) as Swift lacks an `OCMock` like tool.
+
+The `shared` construct in the example is a utility function we provide (in the `Component` base class) that simply returns the same instance every time this `var` is accessed (as opposed to the one above it, which returns a new instance each time).
+
+It's also recommended that you use the component to also construct the `ViewController` that is paired with this component. As you can see in the example above, this allows us to pass in all the dependencies that the `ViewController` needs without the `ViewController` even being aware that you're using a DI system in your project. As noted in the "Benefits of DI" document, it's best to pass in protocols instead of concrete classes or structs.
+
+# Dependencies
+
+If components ended here, they would simply be a container to hold all the "dependencies" of each `ViewController`. This, in itself, is useful as it allows for better unit-tests for your `ViewController` class. Of course, `UIViewController` subclasses are often not easily unit-testable, which is why our RIB architecture (and many others) splits out the "business logic" into a separate class which is easily unit testable.
+
+The real power comes from being able to fetch items from other `Components` as well. 
+
+In order to do this, we specify the items we'd like to fetch from other components in a protocol referred to as a `Dependency Protocol`. At Uber, the dependency protocol associated with a `NameEntryComponent` would be called `NameEntryDependency`. Here is an example (**Note:** We've already used this protocol above in the generic parameter of `Component` class) :
+
+```swift
+protocol LoginDependency: Dependency {
+	var imageCache: ImageCache { get }
+	var networkService: NetworkService { get }
+}
+```
+
+# Using the Component
+
+The nice thing is that you're now ready to write and compile code even though you may not be ready to run the needle command-line code-generator tool. We've also not told the system which `Component` the `imageCache` and `networkService` are supposed to come from.
+
+The example only uses items that are created at the current `Scope`. If we also wanted to pass items into our ViewController which we expect to fetch from other Scopes, then the loginViewController would look like:
+
+```swift
+var loginViewController: UIViewController {
+   return LoginViewController(tokenProvider: tokenProvider, imageCache: dependency.imageCache, button: okButton)
+}
+```
+So even though there may not be any class (yet) that conforms to your `LoginDependency` protocol, you can still (successfully) compile this file by simply assuming there will be.
```

Comment:
Not sure if we want to emphasize that the code is compilable at this point. Technically we want everyone to integrate the generator first before writing any DI code. This would gives us real/more compile-time safety.

Reply:
ok, removed